README.md
# Debit Credit Card Widget

A Flutter widget for displaying debit and credit card designs.

![Example Image](https://github.com/vaibhav13062/debit_credit_card_widget/raw/master/image.png)


## Installation

To use this package, add `debit_credit_card_widget` as a dependency in your `pubspec.yaml` file:
```yaml
dependencies:
  debit_credit_card_widget: ^1.0.2
```

## Getting started
Import the package into your Flutter code:

```dart
import 'package:debit_credit_card_widget/debit_credit_card_widget.dart';
```

## Usage
To create a DebitCreditCardWidget, simply instantiate it with the desired properties:

```dart
DebitCreditCardWidget(
cardHolderName: "John Doe",
cardNumber: "0000000000000000",
cardExpiry: "1224",
// ... other properties
),
```

## Properties

| Param                       | Description                                                                |
|-----------------------------|----------------------------------------------------------------------------|
| `width`                     | Give specific width to the card (take as much space available by default). |
| `cardHolderName`            | Allow you to set the cardholder's name("John Doe" - Default).              |
| `cardNumber`                | Allow you to set the card Number("4 X 0000" - Default)(Max 16).            |
| `cardExpiry`                | Allow you to set the card Expiry("0000" - Default)(Max 4).                 |
| `validFrom`                 | Allow you to set the card Valid From(Max 4).                               |
| `cardDecoration`            | Allows you to customize whole card decoration.                             |
| `textColor`                 | Allows you to change Text Color(White - Default).                          |
| `color1`                    | Allows you to change Gradient First Color(Pink Accent - Default).          |
| `color2`                    | Allows you to change Gradient Second Color(Black - Default).               |
| `showNFC`                   | Allows you to show Hide NFC Logo.                                          |
| `cardBrand`                 | `CardBrand` Enum can be Used to Specify card brand (eg.`CardBrand.visa`).  |
| `cardBrandImage`            | If cardBrand is set to `CardBrand.custom` cardBrandImage will be used.     |
| `cardType`                  | `CardType` Enum can be Used to Specify card brand (eg.`CardType.credit`).  |
| `customCardType`            | If cardType is set to `CardType.custom` customCardType will be used.       |
| `customCompanyImage`        | Allows you to add custom image to top right of the card.                   |
| `backgroundDecorationImage` | Allows you to add background image int he card.                            |

## License

This package is open-source and released under the MIT License.

## Issues and Contributions
Please report any issues or contribute to this project on GitHub.

## Additional information

- Made by :- Vaibhav Chandolia
- Email:- chandolia.vaibhav@gmail.com
- Website:- vaibhavchandolia.info


CHANGELOG.md
## 1.0.2

* Bug fixes


test/debit_credit_card_widget_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:debit_credit_card_widget/debit_credit_card_widget.dart';

void main() {
  testWidgets('MyWidget has a title and message', (tester) async {
    // Create the widget by telling the tester to build it.
    await tester.pumpWidget(const DebitCreditCardWidget());
  });
}


example/lib/main.dart
import 'package:debit_credit_card_widget/debit_credit_card_widget.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'DEBIT/CREDIT CARD EXAMPLE',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'DEBIT/CREDIT CARD EXAMPLE'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: const Center(
        child: DebitCreditCardWidget(
          cardType: CardType
              .debit, // CAN SET CARD TYPE AS DEBIT,CREDIT OR CUSTOM(FOR CUSTOM USE :- customCardType )
          cardBrand: CardBrand
              .mastercard, // CAN SET CARD BRAND (VISA,MASTERCARD,AMERICAN EXPRESS , DISCOVER, RUPAY OR CUSTOM(FOR CUSTOM USE :- cardBrandImage))
          cardExpiry: "2140", // CARD EXPIRY DATE
          color1: Colors.purple, // GRADIENT FIRST COLOR
          showNFC: false, // BOOL TO SHOW OR HIDE NFC LOGO
          cardNumber: "1234123412341234", // CARD NUMBER ONLY 16 CHARACTERS
          cardHolderName: "Vaibhav Chandolia", // CARD HOLDER NAME
          backgroundDecorationImage: DecorationImage(
              image: AssetImage(
                  "assets/images/background.png")), // YOU CAN ADD BACKGROUND IMAGE FROM HERE
        ),
      ),
    );
  }
}


lib/src/debit_credit_card_widget.dart
import 'package:basic_utils/basic_utils.dart';
import 'package:debit_credit_card_widget/src/Enums.dart';
import 'package:flutter/material.dart';

class DebitCreditCardWidget extends StatelessWidget {
  final double? width;
  final String cardHolderName;
  final String cardNumber;
  final String cardExpiry;
  final String? validFrom;
  final BoxDecoration? cardDecoration;
  final Color textColor;
  final Color color2;
  final Color color1;
  final bool showNFC;
  final ImageProvider<Object> cardBrandImage;
  final CardBrand cardBrand;
  final CardType cardType;
  final String customCardType;
  final DecorationImage? backgroundDecorationImage;
  final ImageProvider<Object>? customCompanyImage;
  const DebitCreditCardWidget({
    super.key,
    this.width,
    this.cardHolderName = "John Doe",
    this.cardNumber = "0000000000000000",
    this.cardExpiry = "0000",
    this.cardDecoration,
    this.textColor = Colors.white,
    this.validFrom,
    this.color2 = Colors.black,
    this.color1 = Colors.pinkAccent,
    this.showNFC = true,
    this.cardBrand = CardBrand.visa,
    this.cardBrandImage = const AssetImage(
      "assets/images/visa_logo.png",
      package: "debit_credit_card_widget",
    ),
    this.cardType = CardType.debit,
    this.customCardType = "DEBIT",
    this.customCompanyImage,
    this.backgroundDecorationImage,
  });

  @override
  Widget build(BuildContext context) {
    return SizedBox(
        width: (width ?? MediaQuery.of(context).size.width),
        height: ((width ?? MediaQuery.of(context).size.width)) * 0.56574,
        child: Card(
            margin: const EdgeInsets.all(0),
            clipBehavior: Clip.antiAliasWithSaveLayer,
            elevation: 3,
            shape:
                RoundedRectangleBorder(borderRadius: BorderRadius.circular(15)),
            child: Container(
                decoration: cardDecoration ??
                    BoxDecoration(
                      image: backgroundDecorationImage,
                      gradient: LinearGradient(
                        begin: Alignment.topLeft,
                        end: const Alignment(0.6,
                            1.7), // 10% of the width, so there are ten blinds.
                        colors: [color1, color2], // red to yellow
                        tileMode: TileMode
                            .repeated, // repeats the gradient over the canvas
                      ),
                    ),
                child: Container(
                  margin: const EdgeInsets.only(left: 20, right: 20),
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: <Widget>[
                      const SizedBox(
                        height: 10,
                      ),
                      Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: [
                          Text(getCardTypeString(cardType, customCardType),
                              style: TextStyle(fontSize: 16, color: textColor)),
                          customCompanyImage == null
                              ? Container()
                              : Image(
                                  image: customCompanyImage!,
                                  height: (width ??
                                          MediaQuery.of(context).size.width) /
                                      13,
                                ),
                        ],
                      ),
                      Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: <Widget>[
                          Image(
                            image: const AssetImage(
                                'assets/images/chip_logo.png',
                                package: "debit_credit_card_widget"),
                            height:
                                (width ?? MediaQuery.of(context).size.width) /
                                    11,
                          ),
                          Visibility(
                            visible: showNFC,
                            child: Image(
                                color: textColor,
                                image: const AssetImage(
                                  'assets/images/contactless_logo.png',
                                  package: "debit_credit_card_widget",
                                ),
                                height: (width ??
                                        MediaQuery.of(context).size.width) /
                                    15),
                          ),
                        ],
                      ),
                      Text(
                        StringUtils.addCharAtPosition(
                            cardNumber.toString(), " ", 4,
                            repeat: true),
                        style: TextStyle(
                            fontSize: 25,
                            letterSpacing: 3,
                            fontFamily: "Kredit",
                            package: "debit_credit_card_widget",
                            color: textColor),
                      ),
                      Row(
                        crossAxisAlignment: CrossAxisAlignment.end,
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: <Widget>[
                          Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: <Widget>[
                              Row(
                                children: [
                                  Visibility(
                                    visible: validFrom != null,
                                    child: Row(
                                      children: <Widget>[
                                        Text("VALID\nFROM",
                                            style: TextStyle(
                                                fontSize: 9,
                                                fontWeight: FontWeight.w800,
                                                color: textColor)),
                                        const SizedBox(
                                          width: 10,
                                        ),
                                        Text(
                                            StringUtils.addCharAtPosition(
                                                validFrom.toString(), "/", 2),
                                            style: TextStyle(
                                                fontSize: 22,
                                                letterSpacing: 2,
                                                fontFamily: "Kredit",
                                                package:
                                                    "debit_credit_card_widget",
                                                color: textColor))
                                      ],
                                    ),
                                  ),
                                  Visibility(
                                      visible: validFrom != null,
                                      child: const SizedBox(
                                        width: 50,
                                      )),
                                  Row(
                                    children: <Widget>[
                                      Text("VALID\nTHRU",
                                          style: TextStyle(
                                              fontSize: 9,
                                              fontWeight: FontWeight.w800,
                                              color: textColor)),
                                      const SizedBox(
                                        width: 10,
                                      ),
                                      Text(
                                          StringUtils.addCharAtPosition(
                                              cardExpiry.toString(), "/", 2),
                                          style: TextStyle(
                                              fontSize: 22,
                                              letterSpacing: 2,
                                              fontFamily: "Kredit",
                                              package:
                                                  "debit_credit_card_widget",
                                              color: textColor))
                                    ],
                                  ),
                                ],
                              ),
                              const SizedBox(
                                height: 5,
                              ),
                              Text(
                                cardHolderName,
                                style: TextStyle(
                                    fontSize: 20,
                                    fontFamily: "Kredit",
                                    package: "debit_credit_card_widget",
                                    color: textColor),
                                overflow: TextOverflow.ellipsis,
                              )
                            ],
                          ),
                          Image(
                              image:
                                  getCardBrandImage(cardBrand, cardBrandImage),
                              width: 65)
                        ],
                      ),
                      const SizedBox(
                        height: 1,
                      ),
                    ],
                  ),
                ))));
  }

  String getCardTypeString(CardType cardType, String customCardType) {
    switch (cardType) {
      case CardType.debit:
        return "DEBIT";
      case CardType.credit:
        return "CREDIT";
      case CardType.custom:
        return customCardType;
    }
  }

  ImageProvider<Object> getCardBrandImage(
      CardBrand cartType, ImageProvider<Object> customCardImage) {
    switch (cartType) {
      case CardBrand.mastercard:
        return const AssetImage(
          "assets/images/mastercard_logo.png",
          package: "debit_credit_card_widget",
        );
      case CardBrand.visa:
        return const AssetImage(
          "assets/images/visa_logo.png",
          package: "debit_credit_card_widget",
        );
      case CardBrand.americanExpress:
        return const AssetImage(
          "assets/images/americanExpress_logo.png",
          package: "debit_credit_card_widget",
        );
      case CardBrand.discover:
        return const AssetImage(
          "assets/images/discover_logo.png",
          package: "debit_credit_card_widget",
        );
      case CardBrand.rupay:
        return const AssetImage(
          "assets/images/rupay_logo.png",
          package: "debit_credit_card_widget",
        );
      case CardBrand.custom:
        return customCardImage;
    }
  }
}
